optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -0,0 +1,509 @@
+                minimumConsecutiveWorkingDays(constraintFactory),
+                maximumConsecutiveWorkingDays(constraintFactory)
+        };
+    }
+
+    // ############################################################################
+    // Hard constraints
+    // ############################################################################
+
+    // A nurse can only work one shift per day, i.e. no two shift can be assigned to the same nurse on a day.
+    Constraint oneShiftPerDay(ConstraintFactory constraintFactory) {
+        return constraintFactory
+                .fromUniquePair(ShiftAssignment.class,
+                        Joiners.equal(ShiftAssignment::getEmployee, ShiftAssignment::getShiftDate))
+                .penalize("oneShiftPerDay", HardSoftScore.ONE_HARD);
+    }
+
+    // ############################################################################
+    // Soft constraints
+    // ############################################################################
+    Constraint minimumAndMaximumNumberOfAssignments(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(MinMaxContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine.getContractLineType() == ContractLineType.TOTAL_ASSIGNMENTS &&
+                        minMaxContractLine.isEnabled())
+                .join(Employee.class, Joiners.equal(MinMaxContractLine::getContract, Employee::getContract))
+                .join(ShiftAssignment.class, Joiners.equal((contractLine, employee) -> employee, ShiftAssignment::getEmployee))
+                .groupBy((line, employee, shift) -> ImmutablePair.of(employee, line), ConstraintCollectors.countTri())
+                .filter((employeeContractPair, shiftCount) -> employeeContractPair.getRight().isViolated(shiftCount))
+                .penalize("Minimum and maximum number of assignments", HardSoftScore.ONE_SOFT,
+                        (employeeContractPair, shiftCount) -> employeeContractPair.getRight().getViolationAmount(shiftCount));
+    }
+
+    Constraint minimumConsecutiveWorkingDays(ConstraintFactory constraintFactory) {
+        return getConsecutiveShifts(constraintFactory.from(MinMaxContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine
+                        .getContractLineType() == ContractLineType.CONSECUTIVE_WORKING_DAYS &&
+                        minMaxContractLine.isMinimumEnabled()))
+                                .filter((employeeContractPair,
+                                        shiftList) -> shiftList.getLength() < employeeContractPair.getRight().getMinimumValue())
+                                .penalize("minimumConsecutiveWorkingDays", HardSoftScore.ONE_SOFT,
+                                        (employeeContractPair, shiftList) -> employeeContractPair.getRight()
+                                                .getViolationAmount(shiftList.getLength()));
```

Comment:
I would put the nested stream first, into its own variable. That way, this will become much better to read.
Also, the formatting of the consecutive filters and penalize doesn't seem right here. They should be on the same level.